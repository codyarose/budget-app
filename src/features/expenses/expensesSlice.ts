import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { v4 as uuid } from 'uuid'
import { Moment } from 'moment'

import { RootState } from '../../app/store'
import moment from 'moment'

export interface Expense {
	id: string
	description: string
	note: string
	amount: number
	createdAt: Moment | number
}

export type ExpenseData = Omit<Expense, 'id'>

export const initialState: Expense[] = []

export const expensesSlice = createSlice({
	name: 'expenses',
	initialState,
	reducers: {
		addExpense: (state, action: PayloadAction<Partial<ExpenseData>>) => {
			const {
				description = '',
				note = '',
				amount = 0,
				createdAt = 0,
			} = action.payload
			state.push({
				id: uuid(),
				description,
				note,
				amount,
				createdAt,
			})
		},
		removeExpense: (state, action: PayloadAction<string>) => {
			const id = action.payload
			const index = state.findIndex((expense) => expense.id === id)
			if (index === -1) {
				return state
			} else {
				state.splice(index, 1)
			}
		},
		editExpense: (
			state,
			action: PayloadAction<{
				editedExpense: Partial<ExpenseData>
				id: string
			}>,
		) => {
			const { editedExpense, id } = action.payload
			return state.map((expense) => {
				if (expense.id === id) {
					return {
						...expense,
						...editedExpense,
					}
				} else {
					return expense
				}
			})
		},
	},
})

export const { addExpense, removeExpense, editExpense } = expensesSlice.actions

export const selectExpenses = (state: RootState): Expense[] => state.expenses

export const selectVisibleExpenses = (state: RootState): Expense[] => {
	const expenses = state.expenses
	const { startDate, endDate, text, sortBy } = state.filter

	return expenses
		.filter((expense) => {
			const createdAtMoment = moment(expense.createdAt)
			const momentStartDate = moment(startDate)
			const momentEndDate = moment(endDate)

			const startDateMatch = startDate
				? momentStartDate.isSameOrBefore(createdAtMoment, 'day')
				: true
			const endDateMatch = endDate
				? momentEndDate.isSameOrAfter(createdAtMoment, 'day')
				: true
			const textMatch = expense.description
				.toLowerCase()
				.includes(text.toLowerCase())

			return startDateMatch && endDateMatch && textMatch
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1
			} else {
				return 1
			}
		})
}

export const selectTotal = (state: RootState): number => {
	const expenses = selectVisibleExpenses(state)

	if (!expenses.length) return 0

	return expenses.reduce((acc, expense) => {
		return acc + expense.amount
	}, 0)
}

export default expensesSlice.reducer
