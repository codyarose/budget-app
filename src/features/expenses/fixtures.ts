export default [
	{
		id: '123abc',
		description: 'Title',
		note: '',
		amount: 140500,
		createdAt: 1598763926001,
	},
	{
		id: '456def',
		description: 'Blah',
		note: 'This is a note',
		amount: 2650,
		createdAt: 1598763926009,
	},
	{
		id: '789ghi',
		description: 'Another one',
		note: 'Text goes here',
		amount: 340098,
		createdAt: 1598763926345,
	},
]
