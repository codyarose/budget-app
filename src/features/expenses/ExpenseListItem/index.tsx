import React, { FC } from 'react'
import { Link } from 'react-router-dom'

import { Expense } from '../expensesSlice'

const ExpenseListItem: FC<Expense> = ({
	description,
	amount,
	createdAt,
	note,
	id,
}) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h4>{description}</h4>
			</Link>
			<p>
				{amount} - {createdAt}
			</p>
			{note && <p>{note}</p>}
		</div>
	)
}

export default ExpenseListItem
