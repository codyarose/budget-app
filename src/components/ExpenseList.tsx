import React, { FC } from 'react'
import { selectVisibleExpenses } from '../features/expenses/expensesSlice'
import { useSelector } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'

const ExpenseList: FC = () => {
	const expenses = useSelector(selectVisibleExpenses)
	return (
		<div>
			<h1>Expense List</h1>
			{expenses.map((expense) => (
				<ExpenseListItem key={expense.id} {...expense} />
			))}
		</div>
	)
}

export default ExpenseList
