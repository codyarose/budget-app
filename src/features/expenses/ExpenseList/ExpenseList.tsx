import React, { FC } from 'react'
import ExpenseListItem from '../ExpenseListItem'
import { useSelector } from 'react-redux'
import { selectVisibleExpenses } from '../expensesSlice'

const ExpenseList: FC = () => {
	const expenses = useSelector(selectVisibleExpenses)

	return (
		<div>
			{expenses.length === 0 ? (
				<p>Add an expense</p>
			) : (
				expenses.map((expense) => (
					<ExpenseListItem key={expense.id} {...expense} />
				))
			)}
		</div>
	)
}

export default ExpenseList
