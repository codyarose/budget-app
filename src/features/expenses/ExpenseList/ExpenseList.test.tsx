import React from 'react'
import { shallow } from 'enzyme'

import ExpenseList from './index'
import expenses from '../fixtures'

describe('ExpenseList component', () => {
	it('should render with expenses', () => {
		const wrapper = shallow(<ExpenseList expenses={expenses} />)

		expect(wrapper).toMatchSnapshot()
	})

	it('show render with empty message', () => {
		const wrapper = shallow(<ExpenseList expenses={[]} />)

		expect(wrapper).toMatchSnapshot()
	})
})
