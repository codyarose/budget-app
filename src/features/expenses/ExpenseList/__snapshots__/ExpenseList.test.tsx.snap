// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList component should render with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={140500}
    createdAt={1598763926001}
    description="Title"
    id="123abc"
    key="123abc"
    note=""
    user="user1"
  />
  <ExpenseListItem
    amount={2650}
    createdAt={1598763926009}
    description="Blah"
    id="456def"
    key="456def"
    note="This is a note"
    user="user2"
  />
  <ExpenseListItem
    amount={340098}
    createdAt={1598763926345}
    description="Another one"
    id="789ghi"
    key="789ghi"
    note="Text goes here"
    user="user1"
  />
</div>
`;

exports[`ExpenseList component show render with empty message 1`] = `
<div>
  <p>
    Add an expense
  </p>
</div>
`;
